File: integrations/csharp/otel/OtelSink.cs
using System.Collections.Generic;
using Axora.Logger.Sinks;

namespace Axora.Logger.Otel;

public interface IOtelLoggerLike
{
    void Emit(IDictionary<string, object?> record);
}

public interface IOtelLoggerProviderLike
{
    IOtelLoggerLike GetLogger(string name, string version);
}

public sealed class OtelSink : ISink
{
    private readonly IOtelLoggerLike _logger;

    public OtelSink(IOtelLoggerProviderLike provider)
    {
        _logger = provider.GetLogger("axora-logger", "0.1.0");
    }

    public Task ExportAsync(WideEventPayload @event)
    {
        var attributes = new Dictionary<string, object?>
        {
            ["axora.event_id"] = @event.EventId,
            ["axora.operation"] = @event.Operation,
            ["axora.status"] = @event.Status,
            ["axora.duration_ms"] = @event.DurationMs,
            ["service.name"] = @event.Service,
            ["deployment.environment.name"] = @event.Environment,
        };

        foreach (var (key, value) in @event.Context)
        {
            attributes[$"axora.ctx.{key}"] = value;
        }

        if (@event.Error.Message is not null)
        {
            attributes["exception.type"] = @event.Error.Type;
            attributes["exception.message"] = @event.Error.Message;
            attributes["exception.stacktrace"] = @event.Error.Stack;
        }

        _logger.Emit(
            new Dictionary<string, object?>
            {
                ["severityNumber"] = SeverityNumber(@event.Level),
                ["severityText"] = @event.Level.ToUpperInvariant(),
                ["body"] = @event.Operation,
                ["attributes"] = attributes,
                ["timestamp"] = DateTimeOffset.Parse(@event.TimestampEnd).ToUnixTimeMilliseconds() * 1_000_000,
                ["observedTimestamp"] = DateTimeOffset.Parse(@event.TimestampStart).ToUnixTimeMilliseconds() * 1_000_000,
                ["resource"] = new Dictionary<string, object?>
                {
                    ["attributes"] = new Dictionary<string, object?>
                    {
                        ["service.name"] = @event.Service,
                        ["deployment.environment.name"] = @event.Environment,
                    },
                },
            });

        return Task.CompletedTask;
    }

    private static int SeverityNumber(string level) => level switch
    {
        "info" => 9,
        "warn" => 13,
        "error" => 17,
        "fatal" => 21,
        _ => 9,
    };
}

File: integrations/typescript/otel/src/OtelSink.ts
import type { Sink, WideEventPayload } from '@openakta/logger-core';
import { SeverityNumber } from '@opentelemetry/api-logs';

export interface OtelLoggerLike {
  emit(record: Record<string, unknown>): void;
}

export interface OtelLoggerProviderLike {
  getLogger(name: string, version?: string): OtelLoggerLike;
}

export interface OtelSinkOptions {
  provider: OtelLoggerProviderLike;
  loggerName?: string;
}

const SEVERITY_MAP: Record<WideEventPayload['level'], SeverityNumber> = {
  info: SeverityNumber.INFO,
  warn: SeverityNumber.WARN,
  error: SeverityNumber.ERROR,
  fatal: SeverityNumber.FATAL,
};

export class OtelSink implements Sink {
  private readonly logger: OtelLoggerLike;

  constructor(options: OtelSinkOptions) {
    this.logger = options.provider.getLogger(options.loggerName ?? 'openakta-logger', '0.1.0');
  }

  async export(event: WideEventPayload): Promise<void> {
    const attributes: Record<string, unknown> = {
      'openakta.event_id': event.event_id,
      'openakta.operation': event.operation,
      'openakta.status': event.status,
      'openakta.duration_ms': event.duration_ms,
      'service.name': event.service,
      'deployment.environment.name': event.environment,
    };

    for (const [key, value] of Object.entries(event.context)) {
      attributes[`openakta.ctx.${key}`] = value;
    }

    if (event.error.message) {
      attributes['exception.type'] = event.error.type;
      attributes['exception.message'] = event.error.message;
      attributes['exception.stacktrace'] = event.error.stack;
    }

    this.logger.emit({
      severityNumber: SEVERITY_MAP[event.level] ?? SeverityNumber.INFO,
      severityText: event.level.toUpperCase(),
      body: event.operation,
      attributes,
      timestamp: Date.parse(event.timestamp_end) * 1_000_000,
      observedTimestamp: Date.parse(event.timestamp_start) * 1_000_000,
      resource: {
        attributes: {
          'service.name': event.service,
          'deployment.environment.name': event.environment,
        },
      },
    });
  }
}
